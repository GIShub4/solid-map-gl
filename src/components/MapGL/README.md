# Solid Component for [Mapbox GL JS Map](https://docs.mapbox.com/mapbox-gl-js/api/map/#map-parameters)

## Props

| Prop             | Type    | Description                           |
| ---------------- | ------- | ------------------------------------- |
| class            | object  | css class attached to map element     |
| classlist        | object  | css classlist attached to map element |
| viewport         | object  |                                       |
| options          | object  |                                       |
| triggerResize    | boolean |                                       |
| onMouseMove      | Event   |                                       |
| onViewportChange | Event   |                                       |

_\*required_

## Example

```jsx
import { Component, createSignal } from 'solid-js'
import MapGL, { Viewport } from 'solid-map-gl'

const Map: Component = props => {
    const [viewport, setViewport] = createSignal({
        center: [0, 0]
        zoom: 6,
    } as Viewport)

    return (
    <MapGL
        options={{
            accessToken: { MAPBOX_ACCESS_TOKEN },
            style: 'mapbox://styles/mapbox/light-v10',
        }}
        viewport={viewport()}
        onViewportChange={evt => setViewport(evt)}>
    </MapGL>
    )
}
```
