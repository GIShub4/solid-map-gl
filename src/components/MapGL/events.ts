const events = [
  'onResize',
  'onRemove',
  'onMouseDown',
  'onMouseUp',
  'onMouseOver',
  'onMouseMove',
  'onClick',
  'onDblClick',
  'onMouseEnter',
  'onMouseLeave',
  'onMouseOut',
  'onContextMenu',
  'onTouchStart',
  'onTouchEnd',
  'onTouchMove',
  'onTouchCancel',
  'onMoveStart',
  'onMove',
  'onMoveEnd',
  'onDragStart',
  'onDrag',
  'onDragEnd',
  'onZoomStart',
  'onZoom',
  'onZoomEnd',
  'onRotateStart',
  'onRotate',
  'onRotateEnd',
  'onPitchStart',
  'onPitch',
  'onPitchEnd',
  'onBoxZoomStart',
  'onBoxZoomEnd',
  'onBoxZoomCancel',
  'onWebglContextLost',
  'onWebglContextRestored',
  'onLoad',
  'onError',
  'onData',
  'onStyleData',
  'onSourceData',
  'onDataLoading',
  'onStyleDataLoading',
  'onSourceDataLoading',
]

export { events as default }
