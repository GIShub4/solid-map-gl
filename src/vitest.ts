import '@testing-library/jest-dom'
import 'jsdom-worker'

// @ts-ignore
window.matchMedia =
  window.matchMedia ||
  (() => ({
    matches: false,
    addListener: () => {},
    removeListener: () => {},
  }))
