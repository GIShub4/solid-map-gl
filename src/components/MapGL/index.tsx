import {
  createSignal,
  createEffect,
  onMount,
  onCleanup,
  createContext,
  useContext,
  Component,
  useTransition,
} from 'solid-js'
import { mapEvents } from '../../events'
import mapboxgl from 'mapbox-gl'
import 'mapbox-gl/dist/mapbox-gl.css'
import type MapboxMap from 'mapbox-gl/src/ui/map'
import type MapboxOptions from 'mapbox-gl/src/ui/map'
import type { LngLatLike } from 'mapbox-gl/src/geo/lng_lat.js'
import type { LngLatBounds } from 'mapbox-gl/src/geo/lng_lat_bounds.js'
import type { PaddingOptions } from 'mapbox-gl/src/geo/edge_insets.js'

export type Viewport = {
  center?: LngLatLike
  bounds?: LngLatBounds
  zoom?: number
  pitch?: number
  bearing?: number
  padding?: PaddingOptions
}

const [pending, start] = useTransition()
const [isInternal, setInternal] = createSignal(false)

const MapContext = createContext<MapboxMap>()
const useMap = () => useContext(MapContext)

const MapGL: Component<{
  class?: string
  classList?: {
    [k: string]: boolean | undefined
  }
  viewport?: Viewport
  options?: MapboxOptions
  children?: Element | Element[]
  triggerResize?: boolean
  transitionType?: 'flyTo' | 'easeTo' | 'jumpTo'
  onViewportChange?: (viewport: Viewport) => void
  showTileBoundaries?: boolean
  showTerrainWireframe?: boolean
  showPadding?: boolean
  showCollisionBoxes?: boolean
  showOverdrawInspector?: boolean
  repaint?: boolean
  cursorStyle?: string
}> = props => {
  let map: MapboxMap
  let mapRef: HTMLDivElement

  onMount(() => {
    map = new mapboxgl.Map({
      ...props.options,
      container: mapRef,
      interactive: !!props.onViewportChange,
      bounds: props.viewport.bounds,
      center: props.viewport.center,
      zoom: props.viewport.zoom || null,
      pitch: props.viewport.pitch || null,
      bearing: props.viewport.bearing || null,
      fitBoundsOptions: { padding: props.viewport.padding },
    } as MapboxOptions)
  })

  onCleanup(() => map.remove())

  // Hook up events
  createEffect(() =>
    mapEvents.forEach(item => {
      if (props[item]) {
        const event = item.slice(2).toLowerCase()
        const callback = e => props[item](e)
        map.on(event, callback)
        onCleanup(() => map().off(event, callback))
      }
    })
  )

  // Update debug features
  createEffect(() => (map.showTileBoundaries = props.showTileBoundaries))
  createEffect(() => (map.showTerrainWireframe = props.showTerrainWireframe))
  createEffect(() => (map.showPadding = props.showPadding))
  createEffect(() => (map.showCollisionBoxes = props.showCollisionBoxes))
  createEffect(() => (map.showOverdrawInspector = props.showOverdrawInspector))
  createEffect(() => (map.getCanvas().style.cursor = props.cursorStyle))

  // Update map style
  createEffect(prev => {
    prev !== props.options.style && map.setStyle(props.options.style)
  }, props.options.style)

  // Update projection
  createEffect(prev => {
    prev !== props.options.projection &&
      map.setProjection(props.options.projection)
  }, props.options.projection)

  // Hook up viewport events
  createEffect(() => {
    const callback = event => {
      if (event.originalEvent) {
        setInternal(true)
        props.onViewportChange({
          center: map.getCenter(),
          zoom: map.getZoom(),
          pitch: map.getPitch(),
          bearing: map.getBearing(),
          padding: props.viewport.padding,
          bounds: props.viewport.bounds,
        })
      }
    }
    const callEnd = event => setInternal(false)

    map.on('move', callback).on('moveend', callEnd)
    onCleanup(() => map.off('move', callback).off('moveend', callEnd))
  })

  // Update boundaries
  createEffect(prev => {
    if (props.viewport.bounds != prev)
      props.onViewportChange({
        ...props.viewport,
        ...map.cameraForBounds(props.viewport.bounds, {
          padding: props.viewport.padding,
        }),
      })
    return props.viewport.bounds
  }, props.viewport.bounds)

  // Update Viewport
  createEffect(() => {
    if (isInternal()) return
    switch (props.transitionType) {
      case 'easeTo':
        map.stop().easeTo(props.viewport)
      case 'jumpTo':
        map.stop().jumpTo(props.viewport)
      default:
        map.stop().flyTo(props.viewport)
    }
  })

  let index = 0

  createEffect(() => {
    if (props.triggerResize) {
      index = 0
      window.requestAnimationFrame(function loop() {
        if (index < 15) {
          index++
          window.requestAnimationFrame(loop)
          start(() => map.resize())
        }
      })
    }
  })

  return (
    <MapContext.Provider value={() => map}>
      {props.children}
      <section
        ref={mapRef}
        class={props.class || ''}
        classList={props.classList}
        style={{ height: '100%', width: '100%' }}></section>
    </MapContext.Provider>
  )
}

export { MapGL as default, useMap }
