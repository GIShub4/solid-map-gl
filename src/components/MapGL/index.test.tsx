import { describe, it, expect } from 'vitest'
import { render } from '@solidjs/testing-library'
import MapGL from '../..'

describe('Map', () => {
  it('renders', () => {
    const { container, unmount } = render(() => (
      <MapGL options={{ testMode: true }}></MapGL>
    ))
    expect(container.innerHTML).toMatchSnapshot()
    unmount()
  })

  it('renders with static viewport', () =>
    render(() => <MapGL viewport={{ center: [0, 0], zoom: 5 }}></MapGL>))
})
