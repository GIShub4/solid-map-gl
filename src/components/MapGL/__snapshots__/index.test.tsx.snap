// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Map > renders 1`] = `"<div id=\\"cl-0\\" style=\\"position: absolute; inset: 0; z-index: -1;\\"></div>"`;
